worlds/sc2/mission_order/layout_types.py

```diff
@@ -243,7 +255,126 @@ def idx_rect(self, x: str, y: str, width: str, height: str) -> Union[Set[int], N
+                    target_mission = missions[self.get_grid_index(*target)]
+                    if not target_mission.option_empty:
+                        return (distance, target_mission)
+                else:
+                    # Out of bounds
+                    return None
+            return None
+
+        for (idx, mission) in enumerate(missions):
+            if mission.option_empty:
+                continue
+            point = self.get_grid_coordinates(idx)
+            if self.jump_distance_orthogonal > 1:
+                for direction in Canvas.jumps_orthogonal:
+                    target = raycast(point, direction, self.jump_distance_orthogonal)
+                    if target is not None:
+                        (distance, target_mission) = target
+                        if distance > 1:
+                            # Distance 1 orthogonal jumps already come from the base grid
+                            mission.next.append(target[1])
+            if self.jump_distance_diagonal > 0:
+                for direction in Canvas.jumps_diagonal:
+                    target = raycast(point, direction, self.jump_distance_diagonal)
+                    if target is not None:
+                        (distance, target_mission) = target
+                        if distance == 1:
+                            # Keep distance 1 diagonal slots only if the orthogonal neighbours are empty
+                            x_neighbour = jump(point, (direction[0], 0), 1)
+                            y_neighbour = jump(point, (0, direction[1]), 1)
+                            if (
+                                missions[self.get_grid_index(*x_neighbour)].option_empty and
+                                missions[self.get_grid_index(*y_neighbour)].option_empty
+                            ):
+                                mission.next.append(target_mission)
+                        else:
+                            mission.next.append(target_mission)
+
+        # Pick missions near the original start and end to set as default entrance/exit
+        top_left = self.get_grid_coordinates(0)
+        bottom_right = self.get_grid_coordinates(len(missions) - 1)
+
+        def distance_lambda(point: Tuple[int, int]) -> Callable[[Tuple[int, SC2MOGenMission]], int]:
+            return lambda idx_mission: Grid.euclidean_distance_squared(self.get_grid_coordinates(idx_mission[0]), point)
+        
+        closest_to_top_left = sorted(
+            ((idx, mission) for (idx, mission) in enumerate(missions) if not mission.option_empty),
+            key = distance_lambda(top_left)
+        )
+        closest_to_top_left[0][1].option_entrance = True
+        closest_to_bottom_right = sorted(
+            ((idx, mission) for (idx, mission) in enumerate(missions) if not mission.option_empty),
+            key = distance_lambda(bottom_right)
+        )
+        closest_to_bottom_right[0][1].option_exit = True
```

Comment:
I think it's worth only setting the entrance/exit if the user hasn't done that already. It would save some headache from having to undo a default, and would also simplify the doc a little imo. I was worried about proposing that in the doc for implementation reasons, but it's easily done by just wrapping these two operations in an `if (any(x for x in missions if x.entrance))` kind of statement.

Reply:
It's not that simple. Layouts set themselves up before the user gets to overwrite them and don't get a chance to respond to user changes, to ensure the user has the final say over the layout. I'll change it because it is better in this case, but I don't feel great about enabling this generally.